### PR TITLE
Handles directory creation in cellGameDataCheckCreate2

### PR DESCRIPTION
**Pull request by Inviuz**
fixes not enough space errors on some games
plus a stub returning timedout error causing sceNp loop
for "shaman magic"

### CODE REVIEW

**Comment by digitaldude555** [comment]
Initial D Extreme Stage goes ingame now
screenshot 993
Dead Space goes past intro.
screenshot 986

**Comment by Zangetsu38** [comment]
Initial D it is already Playable ^^

**Comment by digitaldude555** [comment]
Ok, but dead space didn't get beyond menu in the master build last I checked.

**Comment by JohnGodgames** [comment]
**Some of WWE games goes further with this PR
Blades of time goes further
Bionic commando goes further
Dark sector goes further
Everybody's Golf: World Tour goes further**
every

**Comment by Inviuz** [comment]
Added creating param.sfo in folder or rewriting it on game's request.
Needs some more testing.

**Comment by Xcedf** [comment]
Dante's Inferno has this type of error
E {PPU[0x100000e] Thread (Game Data Utility) [0x00ef53d0]} cellGame: cellGameDataCheckCreate2(version=0x0, dirName=“”, errDialog=0x0, funcStat=*0x1168b98, container=4294967295)
E {PPU[0x100000e] Thread (Game Data Utility) [0x00ef53d0]} 'cellGameDataCheckCreate2' failed with 0x8002b604 : CELL_GAMEDATA_ERROR_PARAM

and eternal loading, not sure if this related

**Comment by Inviuz** [comment]
Empty dirName can't really do much about that, send me the full log if possible maybe there's something before

**Comment by Xcedf** [comment]
Alright here it is

**Comment by SakataGintokiYT** [comment]
Dante's Inferno-demo
**[Log:](https://gist.github.com/SakataGintokiYT/3859f43065ee1a491c815cad64dfb31f)**

**Comment by Xcedf** [comment]
here it is
https://mega.nz/#!adZGHRiZ

**Comment by Inviuz** [comment]
Xcedf it needs decryption key, but if it's the same as Sakata's then this part looks suspicious.
```
·W 0:00:01.444916 {PPU[0x1000005] Thread () [0x00cb7bc8]} sceNp: sceNpDrmIsAvailable(k_licensee=*0xcbaca0, drm_path=“/host_root/D:/EMU/PS3/RPCS3/dev_hdd0/game/NPEB90206/USRDIR/BIGFILE0.EDAT”)
·W 0:00:01.444999 {PPU[0x1000005] Thread () [0x00cb7bc8]} sceNp: npDrmIsAvailable(): Found DRM license file at /host_root/D:/EMU/PS3/RPCS3/dev_hdd0/game/NPEB90206/USRDIR/BIGFILE0.EDAT
·W 0:00:01.445007 {PPU[0x1000005] Thread () [0x00cb7bc8]} sceNp: npDrmIsAvailable(): Using k_licensee 0x08b6657bd029cd4af00862826cb780475
·W 0:00:01.445194 {PPU[0x1000005] Thread () [0x00cb7bc8]} sceNp: npDrmIsAvailable(): Can't find RAP file for /host_root/D:/EMU/PS3/RPCS3/dev_hdd0/game/NPEB90206/USRDIR/BIGFILE0.EDAT
·E 0:00:01.445251 {PPU[0x1000005] Thread () [0x00cb7bc8]} LDR: EDAT: D:/EMU/PS3/RPCS3/dev_hdd0/game/NPEB90206/USRDIR/BIGFILE0.EDAT has invalid NPD header or already decrypted.
·W 0:00:01.445480 {PPU[0x1000006] Thread () [0x00cb8748]} cellFs: cellFsOpen(path=“”, flags=0, fd=*0x17082bc, arg=*0xd0083718, size=0x8) -> sys_fs_open()
·W 0:00:01.445493 {PPU[0x1000006] Thread () [0x00cb8748]} sys_fs: sys_fs_open(path=“”, flags=0, fd=*0x17082bc, mode=0, arg=*0xd0083718, size=0x8)
·E 0:00:01.445503 {PPU[0x1000006] Thread () [0x00cb8748]} sys_fs: sys_fs_open(“”) failed: path is invalid
·E 0:00:01.445511 {PPU[0x1000006] Thread () [0x00cb8748]} 'cellFsOpen' failed with 0x80010002 : CELL_EINVAL
```

**Comment by Xcedf** [comment]
I have the .rap file and i mannually decrypted all edat files of the game, i also have the disc version, it has the same result as hdd one, it give the loading intro, can give you it's log
https://mega.nz/#!yUwUVLbD

**Comment by Inviuz** [comment]
@Xcedf i need a decryption key to see your files.

**Comment by Xcedf** [comment]
Sorry
https://mega.nz/#!yUwUVLbD!acc0NHxNStmcCuQcFfCnkLadMvzSzRTNz4dNTii5jfs

**Comment by JohnGodgames** [comment]
https://github.com/RPCS3/rpcs3/commit/c07bae4580fa7f99dd94cee039396d0f54d4ebc4

also 

**Damnation**
`F {PPU[0x1000000] Thread (main_thread) [0x014952a4]} class std::logic_error thrown: fs::file is null`

**Bionic commando**
`F {PPU[0x1000006] Thread (dslGameDataInfo) [0x00e85d78]} class std::logic_error thrown: fs::file is null`

**WWE Games still** 
`....................fs::file is null`

**Comment by RainKikyou** [comment]
PDF2  black screen
qq 20170310105319

**Comment by Inviuz** [comment]
@RainKikyou can you send me the full log? from the latest commit
Wasn't project diva returning an error about files being corrupted and ignoring them?

**Comment by JohnGodgames** [comment]
https://github.com/RPCS3/rpcs3/commit/31f1d9c36a515e2393e7577b9fecaa83d7149ed5

**Bionic commando** - almost ingame

**Damnation** - fs null is gone

**WWE** - fs null is gone

**Comment by RainKikyou** [comment]
@Inviuz Sorry I do not know why, can not work and only get a black screen, try master build get the same result

File: rpcs3/Emu/Cell/Modules/cellGame.cpp [hunk 1]
```diff
@@ -448,23 +445,55 @@ error_code cellGameDataCheckCreate2(ppu_thread& ppu, u32 version, vm::cptr<char>
...
+		{
+			const auto vdir = vfs::get(dir);
+			prm->is_temporary = false;
+
+			//older SDK does not define not settable values, hopefully it doesn't just change some values(overwrite)
+			prm->sfo =
+			{
+				{ "TITLE_ID", psf::string(CELL_GAME_SYSP_TITLEID_SIZE, cbSet->setParam->titleId) },
+				{ "TITLE", psf::string(CELL_GAME_SYSP_TITLE_SIZE, cbSet->setParam->title) },
+				{ "VERSION", psf::string(CELL_GAME_SYSP_VERSION_SIZE, cbSet->setParam->dataVersion) }
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/Modules/cellGame.cpp]: Must be { "PARENTAL_LEVEL", cbSet->setParam->parentalLevel }

File: rpcs3/Emu/Cell/Modules/cellGame.cpp [hunk 2]
```diff
@@ -448,23 +445,55 @@ error_code cellGameDataCheckCreate2(ppu_thread& ppu, u32 version, vm::cptr<char>
...
+			for (u32 i = 0; i < CELL_HDDGAME_SYSP_LANGUAGE_NUM; i++)
+			{
+				prm->sfo.emplace(fmt::format("TITLE_%02d", i), psf::string(CELL_GAME_SYSP_TITLE_SIZE, cbSet->setParam->titleLang[i]));
+			}
+			if (!fs::is_dir(vfs::get(dir)))
+			{
+				cellGame.fatal("directory where param.sfo is to be created does not exist");
+				return CELL_GAME_ERROR_INTERNAL;
+			}
+			(fs::is_file(vdir + "/PARAM.SFO"))? psf::save_object(fs::file(vdir + "/PARAM.SFO", fs::rewrite), prm->sfo) : psf::save_object(fs::file(vdir + "/PARAM.SFO", fs::create + fs::excl + fs::write), prm->sfo);
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/Modules/cellGame.cpp]: This is crazy. Must be enough with fs::rewrite

File: rpcs3/Emu/Cell/Modules/cellGame.cpp [hunk 3]
```diff
@@ -448,23 +445,55 @@ error_code cellGameDataCheckCreate2(ppu_thread& ppu, u32 version, vm::cptr<char>
...
+				{ "TITLE_ID", psf::string(CELL_GAME_SYSP_TITLEID_SIZE, cbSet->setParam->titleId) },
+				{ "TITLE", psf::string(CELL_GAME_SYSP_TITLE_SIZE, cbSet->setParam->title) },
+				{ "VERSION", psf::string(CELL_GAME_SYSP_VERSION_SIZE, cbSet->setParam->dataVersion) }		
+			};
+			prm->sfo.emplace("PARENTAL_LEVEL", 0); // I don't care about age restrictions.
+			for (u32 i = 0; i < CELL_HDDGAME_SYSP_LANGUAGE_NUM; i++)
+			{
+				prm->sfo.emplace(fmt::format("TITLE_%02d", i), psf::string(CELL_GAME_SYSP_TITLE_SIZE, cbSet->setParam->titleLang[i]));
+			}
+			if (!fs::is_dir(vfs::get(dir)))
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/Modules/cellGame.cpp]: vdir

File: rpcs3/Emu/Cell/Modules/cellGame.cpp [hunk 4]
```diff
@@ -405,7 +405,10 @@ error_code cellGameDataCheckCreate2(ppu_thread& ppu, u32 version, vm::cptr<char>
 {
 	cellGame.error("cellGameDataCheckCreate2(version=0x%x, dirName=%s, errDialog=0x%x, funcStat=*0x%x, container=%d)", version, dirName, errDialog, funcStat, container);
 
-	if (version != CELL_GAMEDATA_VERSION_CURRENT || errDialog > 1)
+	//older sdk. it might not care about game type.
+	const auto prm = fxm::make<content_permission>(dirName.get_ptr(), psf::registry{});
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/Modules/cellGame.cpp]: It's not used outside the function,

File: rpcs3/Emu/Cell/Modules/cellGame.cpp (hunk 1 above)
Comment (Inviuz) [inline rpcs3/Emu/Cell/Modules/cellGame.cpp]: if i remember correctly it fails to compile when done this way, but using emplace like in 484 we could put anything we want there

File: rpcs3/Emu/Cell/Modules/cellGame.cpp (hunk 2 above)
Comment (Inviuz) [inline rpcs3/Emu/Cell/Modules/cellGame.cpp]: was throwing some null's earlier, but was unrelated to this. i agree it's a bit excessive

File: rpcs3/Emu/Cell/Modules/cellGame.cpp (hunk 1 above)
Comment (Inviuz) [inline rpcs3/Emu/Cell/Modules/cellGame.cpp]: @Nekotekina newest commit what that change failed on travis

**Pull request merged**